src/core/shape_inference/include/nms_shape_inference.hpp

```diff
@@ -184,24 +185,13 @@ std::vector<TRShape> shape_infer(const Node* op,
     const auto boxes_rank = boxes_shape.rank();
     const auto scores_rank = scores_shape.rank();
 
-    auto out_shape = TRShape{TDim(-1), 3};
+    auto out_shape = TRShape{TDim(dim::inf_bound), 3};
     if (boxes_rank.is_static()) {
-        int64_t max_out_boxes_per_class_val;
-        if (const auto max_out_boxes_per_class = get_input_const_data_as<TRShape, int64_t>(op, 2, ta)) {
-            max_out_boxes_per_class_val = max_out_boxes_per_class->front();
-        } else {
-            max_out_boxes_per_class_val = -1;
-        }
-        const auto& num_boxes = boxes_shape[1];
         auto& selected_boxes = out_shape[0];
-        if (num_boxes.is_static()) {
-            const auto min_selected_boxes =
-                std::min(num_boxes.get_length(), static_cast<V>(max_out_boxes_per_class_val));
-            selected_boxes = static_output ? TDim{min_selected_boxes} : TDim{0, min_selected_boxes};
-        } else if (scores_rank.is_static() && num_boxes.get_max_length() != -1 &&
-                   scores_shape[0].get_max_length() != -1 && scores_shape[1].get_max_length() != -1) {
+        if (const auto max_out_boxes_per_class = get_input_const_data_as<TRShape, int64_t>(op, 2, ta)) {
+            const auto& num_boxes = boxes_shape[1];
             const auto min_selected_boxes =
-                std::min(num_boxes.get_max_length(), static_cast<V>(max_out_boxes_per_class_val));
+                std::min(num_boxes.get_max_length(), static_cast<V>(max_out_boxes_per_class->front()));
```

Comment:
If the `num_boxes` hasn't got upper bound, the `get_max_length` will return `-1`, so it will be taken as minimum even if the `max_out_boxes_per_class` is provided.
Consider test case with `boxes_shape` static rank, fully dynamic dimension for `num_boxes`, with provided max_output_boxes_per_class as Constant input.

Reply:
If `num_box` has `-1` as upper bound the `min_selected_boxes` cannot be evaluated even if `max_out_boxes_per_class ` is constant, because when static shape provided it can be lower or greater than `max_out_boxes_per_class`, take `-1` is correct.
Regarding the test will check and add if required.